libwnbd/wnbd_ioctl.cpp

```diff
@@ -139,6 +144,367 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
+    if (!DiUninstallDevice(0, AdapterDevHandle, DevInfoData, 0, RebootRequired)) {
+        Status = GetLastError();
+        LogError(
+            "Could not remove driver. "
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+    }
+
+    return Status;
+}
+
+// We assume we are using a single WNBD adapter per host
+static DWORD FindWnbdAdapterDevice(HDEVINFO* AdapterDevHandle, SP_DEVINFO_DATA* DevInfoData)
+{
+    CHAR TempBuf[2048];
+    memset(TempBuf, 0, sizeof(TempBuf));
+    BOOL Found = FALSE;
+    DWORD Status = ERROR_FILE_NOT_FOUND;
+
+    *AdapterDevHandle = SetupDiGetClassDevsA(&WNBD_GUID, 0, 0, DIGCF_ALLCLASSES | DIGCF_PRESENT);
+    if (INVALID_HANDLE_VALUE == *AdapterDevHandle) {
+        Status = GetLastError();
+        LogError(
+            "SetupDiGetClassDevs failed. "
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        return Status;
+    }
+
+    for (DWORD I = 0; !Found && SetupDiEnumDeviceInfo(*AdapterDevHandle, I, DevInfoData); I++)
+    {
+        if (!SetupDiGetDeviceRegistryPropertyA(*AdapterDevHandle, DevInfoData, SPDRP_HARDWAREID, 0,
+            (PBYTE)TempBuf, sizeof(TempBuf) - 1, 0)) {
+            continue;
+        }
+
+        if (strstr(TempBuf, WNBD_HARDWAREID) != NULL) {
+            Found = TRUE;
+            break;
+        }
+    }
+
+    if (!Found) {
+        Status = GetLastError();
+        if (ERROR_NO_MORE_ITEMS != Status) {
+            LogError(
+                "Failed to locate device with hardware id: %s. Error: %d. Error message: %s",
+                WNBD_HARDWAREID, Status, win32_strerror(Status).c_str());
+        }
+        return Status;
+    }
+
+    return ERROR_SUCCESS;
+}
+
+DWORD RemoveAllDevices(BOOLEAN Force)
+{
+    DWORD BufferSize = 0;
+    DWORD Status = 0;
+    PWNBD_CONNECTION_LIST ConnList = NULL;
+
+    WNBD_OPTION_VALUE OptValue = { WnbdOptBool };
+    /* Disallow new mappings so we can remove all current mappings */
+    OptValue.Data.AsBool = FALSE;
+    Status = WnbdSetDrvOpt("NewMappingsAllowed", &OptValue, FALSE);
+    if (Status) {
+        goto exit;
+    }
+
+    Status = WnbdList(ConnList, &BufferSize);
+    if (!BufferSize) {
+        goto exit;
+    }
+    ConnList = (PWNBD_CONNECTION_LIST)calloc(1, BufferSize);
+    if (NULL == ConnList) {
+        Status = ERROR_NOT_ENOUGH_MEMORY;
+        LogError(
+            "Failed to allocate %d bytes.", BufferSize);
+        goto exit;
+    }
+    Status = WnbdList(ConnList, &BufferSize);
+    if (Status) {
+        goto exit;
+    }
+    CHAR* InstanceName;
+    for (ULONG index = 0; index < ConnList->Count; index++) {
+        InstanceName = ConnList->Connections[index].Properties.InstanceName;
+        WNBD_REMOVE_OPTIONS RemoveOptions = { 0 };
+        RemoveOptions.Flags.HardRemoveFallback = TRUE;
+        RemoveOptions.SoftRemoveTimeoutMs = 30 * 1000;
```

Comment:
This can be an issue: soft removing all disks sequentially with a 30s timeout. For just 20 in-use volumes, you might have to wait 10 minutes.

FWIW, in ceph I'm using a thread pool and a global timeout: https://github.com/petrutlucian94/ceph/blob/windows.12/src/tools/rbd_wnbd/rbd_wnbd.cc#L462-L523

I think we can move that ceph code to libwnbd, maybe adding WnbdRemoveAll with an optional OwnerName argument. Until then, it's probably better to stick to a force removal.

Reply:
Thanks for pointing that out. I added a TODO for it.